rfcs/0100-sparestir-dialect.md

```diff
@@ -0,0 +1,658 @@
+[guide-level-explaination]: #guide-level-explaination
+
+We have the following design goals of SparseTIR:
+
+- G0: SparseTIR is consistent with TVM's ecosystem, which means other components of TVM stack (Relax/Relay/TOPI)
+can interact with SparseTIR smoothly, and enjoy of the benefits of SparseTIR.
+- G1: SparseTIR is expressive, which means we can express most sparse operators in Deep Learning with SparseTIR.
+- G2: SparseTIR is performant, which means we can cover optimizations used in Sparse CPU/GPU libraries.
+
+We will outline the detailed design in the next section.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section outlines the design of SparseTIR, and its interaction with existing components in TVM.
+
+## D0: Programming Interface
+
+A generic SparseTIR program looks like the following, the workload is Sampled-Dense-Dense-Matrix-Multiplication (SDDMM):
+
+```python
+@T.prim_func
+def sddmm(
+    a: T.handle,
+    b: T.handle,
+    x: T.handle,
+    y: T.handle,
+    indptr: T.handle,
+    indices: T.handle,
+    m: T.int32,
+    n: T.int32,
+    feat_size: T.int32,
+    nnz: T.int32,
+) -> None:
+    T.func_attr({"global_symbol": "main", "tir.noalias": True, "sparse_tir_level": 2})
+    # sparse axes
+    I = T.dense_fixed(m)
+    J = T.sparse_variable(I, (n, nnz), (indptr, indices), "int32")
+    J_detach = T.dense_fixed(n)
+    K = T.dense_fixed(feat_size)
+    # sparse buffers
+    A = T.match_sparse_buffer(a, (I, K), "float32")
+    B = T.match_sparse_buffer(b, (J_detach, K), "float32")
+    X = T.match_sparse_buffer(x, (I, J), "float32")
+    Y = T.match_sparse_buffer(y, (I, J), "float32")
+    # sparse iterations
+    with T.sp_iter([I, J, K], "SSR", "sddmm") as [i, j, k]:
+        with T.init():
+            Y[i, j] = 0.0
+        Y[i, j] = Y[i, j] + A[i, k] * B[j, k] * X[i, j]
+```
+
+where we have constructs like **sparse axes**,  **sparse buffers** and **sparse iterations**.
+
+### Sparse Axis
+Sparse axis is a generation of per-dimensional level formats in TACO where we annotate each dimension of a format as **dense**/**sparse** (this dimension is stored in dense or compressed storage) and **fixed**/**variable** (this dimension's extent is fixed or variable). For **sparse**/**variable** axes, we need to specify its dependent axis.
+
+- For axes that are **sparse**, we need to specify a `indices` array to store the column indices.
+- For axes that are **variable**, we need to specify an `indptr` (short for indices pointer) array to store the start offset of each row because the row length is variable and we cannot simply compute element offset with an affine map of indices.
+- An axes that is both **sparse** and **variable** need to be specified with both **indices** and **indptr** array.
+
+```python
+I = T.dense_fixed(m)
+# J1 is a sparse fixed axis, whose dependent axis is I
+# it has maximum length n and number of non-zero elements per row: c,
```

Comment:
Because `indices_1` is a handle instead of buffer, and we need to use the parameter `c` to materialize the `J1_indices` buffer in `lower_sparse_iter` pass, see step 1 in section 3.3.1 of [our paper](https://dl.acm.org/doi/pdf/10.1145/3582016.3582047)

Reply:
Ah, I see.  I was expected the `T.match_sparse_buffer` to be the source of definition for buffers, similar to how `T.match_buffer` defines buffer objects in the `PrimFuncNode::buffer_map`.

Follow-up question, since the `indices_1` handle needs to be annotated with a length `c` and underlying dtype `idtype`, what is the advantage of maintaining a handle until the later lowering pass?  It seems like the extra information is the same information as contained in a `tir::Buffer` object, just packed into a different format.

Reply:
It's just because the `indices_1` is user provided pointers.
In the futher we might accept more advanced sparse tensor inputs in TIR script, but currently we still expect user to provide pointer to `indptr`/`indices` buffer of each dimension as input.